### PR TITLE
Better type hints for `LightningModule` steps

### LINKED ISSUE #17786: LightningModule: predict_step has different signature than other steps
### Outline & Motivation

In the documentation, `(train|validation|test|predict)_step` all accept the same arguments. However, they have very different function signatures. This makes it confusing to subclass as I need to be careful about using different type signatures for each step. Some consistency would be nice.

### Pitch

All steps should be consistent. Either:
```python
def *_step(self, *args: Any, **kwargs: Any) -> Optional[STEP_OUTPUT]:
```
or:
```python
def *_step(self, batch: Any, batch_idx: int, dataloader_idx: int = 0) -> Any:
```
Personally I prefer the latter, as it makes it more clear what arguments will be passed and what types they can be.

### Additional context

_No response_

cc @justusschock @awaelchli @carmocca @borda

### PR DESCRIPTION
**Pull request by adamjstewart**
## What does this PR do?

Fixes #17786

All LightningModule steps have 3 possible function signatures. This PR allows these function signatures to be tested by mypy, and uses consistent type hints and documentation for all steps.

<details>
  <summary><b>Before submitting</b></summary>

- [x] Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

@carmocca

### CODE REVIEW

File: src/lightning/pytorch/core/module.py [hunk 1]
```diff
@@ -648,21 +649,28 @@ def forward(self, *args: Any, **kwargs: Any) -> Any:
...
+    @overload
+    def training_step(self, dataloader_iter: Iterable[Dataloader[Any]]) -> Optional[STEP_OUTPUT]:
+        ...
+
+    def training_step(self, *args: Any, **kwargs: Any) -> Optional[STEP_OUTPUT]:
         r"""Here you compute and return the training loss and some additional metrics for e.g. the progress bar or
         logger.
 
         Args:
-            batch (:class:`~torch.Tensor` | (:class:`~torch.Tensor`, ...) | [:class:`~torch.Tensor`, ...]):
```
Comment (adamjstewart) [inline src/lightning/pytorch/core/module.py:656]: `DataLoader` can output any value, not just `Tensor | tuple[Tensor, ...] | list[Tensor]`. In TorchGeo, we output `dict[str, Tensor]`.

File: src/lightning/pytorch/core/module.py [hunk 2]
```diff
@@ -852,6 +885,18 @@ def predict_step(self, batch: Any, batch_idx: int, dataloader_idx: int = 0) -> A
...
+            batch: The output of your :class:`~torch.utils.data.DataLoader`.
+            batch_idx: The index of this batch.
+            dataloader_idx: The index of the dataloader that produced this batch.
+                (only if multiple dataloaders used)
+
+        Return:
+            - :class:`~torch.Tensor` - The loss tensor
+            - ``dict`` - A dictionary. Can include any keys, but must include the key ``'loss'``.
+            - ``None`` - Skip to the next batch. This is only for automatic optimization.
+                This is not supported for multi-GPU, TPU, IPU, or DeepSpeed.
```
Comment (adamjstewart) [inline src/lightning/pytorch/core/module.py:898]: Not sure what possible outputs to expect here. In TorchGeo, we'll likely want to return `dict[str, Tensor]`. So maybe `Optional[STEP_OUTPUT]` and just say "Predicted output"?

File: src/lightning/pytorch/core/module.py [hunk 3]
```diff
@@ -710,11 +725,13 @@ def validation_step(self, *args: Any, **kwargs: Any) -> Optional[STEP_OUTPUT]:
...
             batch_idx: The index of this batch.
             dataloader_idx: The index of the dataloader that produced this batch.
-                (only if multiple val dataloaders used)
+                (only if multiple dataloaders used)
 
         Return:
-            - Any object or value
-            - ``None`` - Validation will skip to the next batch
+            - :class:`~torch.Tensor` - The loss tensor
+            - ``dict`` - A dictionary. Can include any keys, but must include the key ``'loss'``.
```
Comment (adamjstewart) [inline src/lightning/pytorch/core/module.py:732]: Maybe it doesn't make sense for val/test/predict to possibly return a loss tensor?

File: src/lightning/pytorch/core/module.py [hunk 4]
```diff
@@ -648,21 +650,32 @@ def forward(self, *args: Any, **kwargs: Any) -> Any:
         """
         return super().forward(*args, **kwargs)
 
-    def training_step(self, *args: Any, **kwargs: Any) -> STEP_OUTPUT:  # type: ignore[return-value]
```
Comment (adamjstewart) [inline src/lightning/pytorch/core/module.py:651]: Should train/predict have a required output?

**Comment by adamjstewart** [comment]
It doesn't seem like what I'm trying to do is working for some reason. Might give up and do a simpler fix unless someone has any ideas.

**Review by awaelchli — COMMENTED** [review]
This looks like a great improvement. On master, signature auto-complete (e.g. in pycharm) would generate:

```py
    def training_step(self, *args: Any, **kwargs: Any) -> STEP_OUTPUT:
```

which then requires the user to replace the args kwargs with the batch and batch_idx parameters manually. 

On this PR however, we get

```py
    def training_step(self, batch: Any, batch_idx: int) -> Optional[STEP_OUTPUT]:
```
which is more like what the user needs 99% of the time.

File: src/lightning/pytorch/core/module.py [hunk 5]
```diff
@@ -648,21 +650,32 @@ def forward(self, *args: Any, **kwargs: Any) -> Any:
...
+    @overload
+    def training_step(self, batch: Any, batch_idx: int) -> Optional[STEP_OUTPUT]:
```
Comment (awaelchli) [inline src/lightning/pytorch/core/module.py:654]: Could we make the Optional as part of the `STEP_OUTPUT` like before? This would make it easier to delete. For users who use auto-complete, I believe most of them want no type hints.

File: src/lightning/pytorch/core/module.py [hunk 6]
```diff
@@ -648,21 +650,32 @@ def forward(self, *args: Any, **kwargs: Any) -> Any:
...
+                (only if multiple dataloaders used)
 
         Return:
-            Any of.
-
             - :class:`~torch.Tensor` - The loss tensor
-            - ``dict`` - A dictionary. Can include any keys, but must include the key ``'loss'``
-            - ``None`` - Training will skip to the next batch. This is only for automatic optimization.
+            - ``dict`` - A dictionary. Can include any keys, but must include the key ``'loss'``.
+            - ``None`` - Skip to the next batch. This is only for automatic optimization.
```
Comment (awaelchli) [inline src/lightning/pytorch/core/module.py:678]:
```suggestion
            - ``None`` - Skip to the next batch. This is only supported for automatic optimization.
```
(while we are at it)

File: src/lightning/pytorch/core/module.py (hunk 3 above)
Comment (awaelchli) [reply src/lightning/pytorch/core/module.py:742]: Yes, there is no longer any special functionality for val/test-steps returning an object.

File: src/lightning/pytorch/core/module.py [hunk 7]
```diff
@@ -710,11 +735,13 @@ def validation_step(self, *args: Any, **kwargs: Any) -> Optional[STEP_OUTPUT]:
...
             dataloader_idx: The index of the dataloader that produced this batch.
-                (only if multiple val dataloaders used)
+                (only if multiple dataloaders used)
 
         Return:
-            - Any object or value
-            - ``None`` - Validation will skip to the next batch
+            - :class:`~torch.Tensor` - The loss tensor
+            - ``dict`` - A dictionary. Can include any keys, but must include the key ``'loss'``.
+            - ``None`` - Skip to the next batch. This is only for automatic optimization.
```
Comment (awaelchli) [inline src/lightning/pytorch/core/module.py:743]:
```suggestion
            - ``None`` - Skip to the next batch. This is only supported for automatic optimization.
```

File: src/lightning/pytorch/core/module.py [hunk 8]
```diff
@@ -769,21 +796,33 @@ def validation_step(self, batch, batch_idx, dataloader_idx=0):
...
+                (only if multiple dataloaders used)
 
         Return:
-           Any of.
-
-            - Any object or value
-            - ``None`` - Testing will skip to the next batch
+            - :class:`~torch.Tensor` - The loss tensor
+            - ``dict`` - A dictionary. Can include any keys, but must include the key ``'loss'``.
+            - ``None`` - Skip to the next batch. This is only for automatic optimization.
```
Comment (awaelchli) [inline src/lightning/pytorch/core/module.py:824]:
```suggestion
            - ``None`` - Skip to the next batch. This is only supported for automatic optimization.
```

File: src/lightning/pytorch/core/module.py (hunk 2 above)
Comment (awaelchli) [reply src/lightning/pytorch/core/module.py:916]: Yep, could be anything

**Review by awaelchli — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

File: src/lightning/pytorch/core/module.py (hunk 5 above)
Comment (adamjstewart) [reply src/lightning/pytorch/core/module.py:654]: You mean modify the definition of `STEP_OUTPUT` in `src/lightning/pytorch/utilities/types.py`?

File: src/lightning/pytorch/core/module.py (hunk 3 above)
Comment (adamjstewart) [reply src/lightning/pytorch/core/module.py:742]: In that case should we change the output type to None?

Comment (awaelchli) [reply src/lightning/pytorch/core/module.py:742]: Yes that or maybe `Any | None`. In theory the user can use the LightningModule also standalone (e.g. with Fabric) and returning something there could maybe make sense. @carmocca confirm? :)

Comment (adamjstewart) [reply src/lightning/pytorch/core/module.py:742]: I believe `Any | None` is equivalent to `Any`

**Review by Borda — APPROVED** [review]

**Comment by adamjstewart** [comment]
mypy tests are still failing. The `lr_finder.py` thing is easy to fix, but the `boring_classes.py` issue is more fundamental and more important. I don't think that overrides with a variable number of arguments are supported in Python/mypy. I don't see any easy way around this issue other than using `*args, **kwargs` for all steps and giving up on overrides, or backwards incompatible changes that make all implementations accept exactly 3 args, or splitting the different versions into different methods. I'm guessing the latter isn't acceptable. Should we go with the former?

**Comment by awaelchli** [comment]
Yes, backward incompatible changes on the LightningModule wouldn't be acceptable. The best person to ask for advice on the rest is @carmocca :)

**Comment by carmocca** [comment]
The current code uses `*args, **kwargs` for those limitations that you encountered @adamjstewart. The other suggestions would require introducing new APIs or breaking changes so they wouldn't be justified simply for stricter typing.

My suggestion would be to drop using overloads and just land the other typing improvements. Sorry that I don't have a better solution for you

**Review by awaelchli — APPROVED** [review]

File: src/lightning/pytorch/core/module.py [hunk 9]
```diff
@@ -863,16 +871,8 @@ def predict_step(self, batch, batch_idx, dataloader_idx=0):
...
-        Args:
-            batch: Current batch.
-            batch_idx: Index of current batch.
-            dataloader_idx: Index of the current dataloader.
-
-        Return:
-            Predicted output
         """
+        batch = args[0]
         return self(batch)
```
[the remark this answers is not in the document]
Comment (adamjstewart) [reply src/lightning/pytorch/core/module.py:876]: This PR is designed to fix #17786. If you disagree with this change, I'll likely just close this PR.

[the remark this answers is not in the document]
Comment (awaelchli) [reply src/lightning/pytorch/core/module.py:876]: Couldn't we handle it like so?
```py
batch = kwargs.get("batch", args[0])
```

[the remark this answers is not in the document]
Comment (adamjstewart) [reply src/lightning/pytorch/core/module.py:876]: We could I guess if that's what you want me to do. I'm not sure how common that design pattern is.

File: src/lightning/pytorch/core/module.py [hunk 10]
```diff
@@ -852,6 +851,15 @@ def predict_step(self, batch: Any, batch_idx: int, dataloader_idx: int = 0) -> A
         based accelerator. This happens for ``Trainer(strategy="ddp_spawn")``
         or training on 8 TPU cores with ``Trainer(accelerator="tpu", devices=8)`` as predictions won't be returned.
 
+        Args:
+            batch: The output of your :class:`~torch.utils.data.DataLoader`.
```
[the remark this answers is not in the document]
Comment (carmocca) [reply src/lightning/pytorch/core/module.py:855]: The output of your data iterable, normally a :class:`~torch.utils.data.DataLoader`.

Can you update this in the other places too? Thank you

**Review by carmocca — APPROVED** [review]
